### ISSUE TITLE
problem with messages in node.send

### ISSUE THREAD

**Issue by andysc** [node-red/node-red#1214]
If you use node.send from inside a function node, it doesn't seem to copy the emitted msg properly.
In the flow below, you get 1, then 9 10s, rather than the ascending numbers. 
If you take the delay node out, it works fine, so it's something about references to objects rather than copies of objects, I suspect.

I can work around it with node.send({payload: i}), but I'd prefer not to.

`[{"id":"3935632e.f2d07c","type":"inject","z":"3eed488a.dd7fc","name":"go","topic":"","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"x":130,"y":180,"wires":[["2642f404.f305ac"]]},{"id":"2642f404.f305ac","type":"function","z":"3eed488a.dd7fc","name":"cycle","func":"for (var i=1; i<=10; i++)\n{\n    msg.payload = i;\n    node.send(msg);\n}\n","outputs":1,"noerr":0,"x":310,"y":180,"wires":[["6a81f90e.931578"]]},{"id":"6a81f90e.931578","type":"delay","z":"3eed488a.dd7fc","name":"","pauseType":"rate","timeout":"5","timeoutUnits":"seconds","rate":"2","nbRateUnits":"1","rateUnits":"second","randomFirst":"1","randomLast":"5","randomUnits":"seconds","drop":false,"x":450,"y":240,"wires":[["fb0663e4.764f08"]]},{"id":"fb0663e4.764f08","type":"debug","z":"3eed488a.dd7fc","name":"","active":true,"console":"false","complete":"false","x":530,"y":300,"wires":[]}]`

**Comment by knolleary** [comment]
If you are sending the same message object repeatedly this is what you get.

If you want to reuse the message object multiple times you are responsible for cloning it with:

    var newMessage = RED.utils.cloneMessage(msg);

Will update the docs to make this clearer.

**Comment by andysc** [comment]
but if you take out the delay node, it works fine - you get the 1-10 values as expected. So it seems odd you get different behaviour when you add an extra node later

**Comment by hardillb** [comment]
no, the message that flows down stream is normally the same message and can be updated by anything that holds a reference.

**Comment by andysc** [comment]
that as may be, but users wouldn't have any reason to be aware of that... is there anywhere else that this might be observed?

**Comment by knolleary** [comment]
We can discuss what behaviour is expected or otherwise, but here are the facts of the current behaviour which explains what you observe.

1. the runtime tries to avoid cloning messages where possible as it is an expensive operation. If a node calls `send` with a single message object and is wired to just one other node, it will skip automatically cloning the message as, in theory, there is only the single copy to worry about.

2. the runtime has no way of knowing that a subsequent call to `send` is with a reused message object.

3. the message passing is 'depth-first' up to the point an async operation is encountered.

4. in a flow consisting of `Inject -> Function -> Debug`, the call to `send(msg)` from the Function node triggers the Debug node which displays the message. The call stack then unwinds and allows the Function node to make its next call to `send`. Because the original `msg` has already been dealt with by the Debug node, the fact the msg has been modified is not noticed.

5. In a flow consisting of `Inject -> Function -> Delay -> Debug`, the Delay node contains an async action. This means the Function node is able to make multiple calls to `send` - queuing up all of its messages at the Delay node. They then, asynchronously, arrive at the Debug node where you discover what your thought were multiple individual messages were in fact multiple references to the same message object.


### So what do we do about it?

We've always said that a node is responsible for cloning a message object if they intend to reuse it multiple times. That is something that is probably more explicit in the nodes we publish than in the docs (which is why I marked this as a docs issue in my original response).

That said, it will be an uphill struggle to explain this nuance of message passing to all users - especially those who don't read the docs.

### In summary:

 - we avoid cloning in order to improve performance on the most common flow case (one node, one message, one recipient).
 - users who have wanted to implement the edge case of sending multiple messages (and reuse the object) have either diligently cloned their message or not.
 - if we add cloning to all messages it will penalise all users - and those that have been diligent doubly so as their code will clone the message and then so will ours.

### In conclusion:

 - 🤔

**Comment by andysc** [comment]
thank you for that detailed explanation :) IMHO, FWIW, not many people use node.send, and I would posit that those that do would expect the behaviour that I did, i.e. that distinct messages would be emitted. 
So I think if you cloned the message on node.send it would only impact performance for people who use node.send, would avoid having to explain quite a subtle nuance of NR internals, and would make it behave (I argue) "as expected". 
For the case where there is only one following node (e.g. a debug), the optimisation of not cloning the message then is fine (for the reason you explained).
For people who have gone to the trouble of creating a new message object first (and that really wouldn't have occurred to me, even though it's what I implicitly did when I did node.send({payload: i}) ), then I would argue it's a small hit.

**Comment by andysc** [comment]
another thought ... is there any way for an async node like delay to notice that it's getting repeated pointers to the same message object, and take steps to clone them itself in that situation, so the queue is full of unique objects rather than repeated pointers to the same object? 
I think I'd prefer to see the function node do it in a way which gives the intended outcome rather than rely on all implementors of async nodes to do something special on their side, though.

**Comment by cinhcet** [comment]
I wonder how many contrib-nodes are out there which are not aware of this issue....

**Comment by nielsnl68** [comment]
Maybe i'm a bit late in thid discusion, but @andysc i dont agree with you that the cloning sould be done inside node.send () sorry. I use it sometime with new created messages. With your request that would mean it will always be cloned. An other issue i can think of is that you want to modify the received msg for other nodes to get as well.

The best way, as suggested earlier is to clone the msg yourself when you need that.
Or at least create a second function in node the node object like clonedsend() or something like that.

**Comment by dceejay** [comment]
I kind of can see it both ways... 

On the one hand - In a typical function you just `return msg;` at the end, and that will do it without cloning, and you can return and array of an array of msgs if you want to return multiple messages. If you "have to" also do a `node.send(` within a function then that is either an extra send or as it's part of an async action, in which cases it's "safer" (for all the user assumptions mentioned) to clone it anyway.

On the other hand having an explicit node.sendClone() function would be clear, and wouldn't change existing behaviour - but users would need to know to use it.

**Comment by andysc** [comment]
my biggest worry is the unintended consequences... i.e unexpectedly your code doesn't work as expected (like I saw when I inserted a delay node before the debug node). That doesn't seem to be the Node-RED way to me

**Comment by guillaume-d** [comment]
On a related note I guess one should not expect objects handed out to output nodes to remain intact.

We used to inadvertently expect that, but something in the mqtt-out node or downstream seems to JSON.stringify the payload of the handed-out MQTT message in-place.
Plase note I am no saying this is a bug in Node-RED or MQTT.js: I think when handed over a message does not belong to the user anymore.

So again: users should make copies of messages after sending if they need them!

**Comment by styltsars96** [comment]
For very small cases in scope, that are important, sending a copy of the msg object is actually the key.
`node.send(Object.assign({},msg));` actually works fine, just try not to overdo it.

**Comment by knolleary** [comment]
@styltsars96 we do provide `RED.utils.cloneMessage(msg)` for this purpose. It deals with the edge case where the HTTP In/Response nodes are involved as they send the `msg.req` and `msg.res` properties that are not cloneable.

**Comment by styltsars96** [comment]
@knolleary It is not exposed to function nodes however. The way I understood it, `RED.utils.cloneMessage(msg)`  is only good when we create new nodes. 
Do I need to add the `require('node-red');` in the function global context, through the settings.js to expose it to functions? Or is there a simpler way that eludes me? It could be usefull in the future.
The solution I posted works for the stuff I'm currently working on however. 
Thanks, anyway!

**Comment by knolleary** [comment]
@styltsars96 apologies for the typo - `RED.util.cloneMessage(msg)` (not `utils`) is certainly available to the Function node.

**Comment by styltsars96** [comment]
@knolleary Thank you, just tried it, and it does work. Since it clones internal objects like `res` and `req`, I will be using this from now on.

**Comment by knolleary** [comment]
Since 1.0, node.send will clone all messages it is given - which I think eliminates the original issue being discussed.
